File: file-translation-demo/csharp/FileTranslationDemo/FileTranslateFlow.cs
﻿using common;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Timers;

namespace FileTranslationDemo
{
    public class FileTranslateFlow
    {        
        private FileTranslateClient client;
        private System.Timers.Timer aTimer = null;
        private RozettaApiUser rozettaApiUser;
        private string translateId;
        private bool done = false;
        private const int Interval = 10000;
        public FileTranslateFlow(string baseUrl)
        {            
            client = new FileTranslateClient(baseUrl);
        }
        public async Task<bool> RozettaApiFlowAsync(RozettaApiUser rozettaApiUser, string[] files, string[] langs, int fieldId)
        {
            this.rozettaApiUser = rozettaApiUser;

            long timeStamp = DateTimeOffset.Now.ToUnixTimeMilliseconds();

            FileTranslationOption option = new FileTranslationOption
            {
                Langs = langs,
                AccessKey = rozettaApiUser.AccessKey,
                SecretKey = rozettaApiUser.SecretKey,
                Nonce = timeStamp.ToString(),
                FieldId = fieldId
            };

            translateId = await client.TranslateAsync(files, rozettaApiUser.ContractId, option);

            if (aTimer != null)
                StopTimer(aTimer);

            aTimer = new System.Timers.Timer();

            aTimer.Interval = Interval;

            aTimer.Elapsed += OnTimedEvent;

            aTimer.AutoReset = true;

            aTimer.Enabled = true;

            while (!done)
            {
                Thread.Sleep(1000);
            }

            return true;
        }

        private void OnTimedEvent(Object source, ElapsedEventArgs e)
        {
            // waiting for all translation done
            Task<Translate> task = client.GetOneHistoryAsync(rozettaApiUser, translateId);

            task.Wait(Interval);

            Translate translate = task.Result;

            if (translate==null)
                return;

            foreach (TranslateItem item in translate.items)
            {
                if (!item.done)
                {
                    return;
                }
            }
            List<string> ids = new List<string>();

            foreach (TranslateItem item in translate.items)
            {
                ids.Add(item.translateItemId);
            }

            string strIds = JsonConvert.SerializeObject(ids);

            Task<bool> downloadTask = client.DownloadZipAsync(rozettaApiUser, strIds, @"C:\mydocuments");

            downloadTask.Wait(Interval);

            done = downloadTask.Result;

            if(done)
                StopTimer(aTimer);
        }
        private void StopTimer(System.Timers.Timer timer)
        {
            aTimer.Enabled = false;
            aTimer.Stop();
            aTimer.Dispose();
        }
    }
}

File: file-translation-demo/vb.net/FileTranslationDemo/FileTranslateFlow.vb
Imports common.common

Namespace FileTranslationDemo
    Public Class FileTranslateFlow
        Private client As FileTranslateClient
        Private aTimer As Timers.Timer = DirectCast(Nothing, Timers.Timer)
        Private rozettaApiUser As RozettaApiUser
        Private translateId As String
        Private done As Boolean = False
        Private Const Interval As Integer = 10000

        Public Sub New(ByVal baseUrl As String)
            Me.client = New FileTranslateClient(baseUrl)
        End Sub

        Public Async Function RozettaApiFlowAsync(ByVal rozettaApiUser As RozettaApiUser, ByVal files As String(), ByVal langs As String(), ByVal fieldId As Integer) As Task(Of Boolean)
            Me.rozettaApiUser = rozettaApiUser
            Dim timeStamp As Long = DateTimeOffset.Now.ToUnixTimeMilliseconds()
            Dim [option] As FileTranslationOption = New FileTranslationOption With {
                .Langs = langs,
                .AccessKey = rozettaApiUser.AccessKey,
                .SecretKey = rozettaApiUser.SecretKey,
                .ContractId = rozettaApiUser.FileContractId,
                .Nonce = timeStamp.ToString(),
                .FieldId = fieldId
            }
            Me.translateId = Await Me.client.TranslateAsync(files, [option])
            If Me.aTimer IsNot DirectCast(Nothing, Object) Then Me.StopTimer(Me.aTimer)
            Me.aTimer = New Timers.Timer()
            Me.aTimer.Interval = Interval
            AddHandler Me.aTimer.Elapsed, AddressOf Me.OnTimedEvent
            Me.aTimer.AutoReset = True
            Me.aTimer.Enabled = True

            While Not Me.done
                Threading.Thread.Sleep(1000)
            End While

            Return True
        End Function

        Private Sub OnTimedEvent(ByVal source As Object, ByVal e As Timers.ElapsedEventArgs)
            ' waiting for all translation done
            Dim task As Task(Of Translate) = Me.client.GetOneHistoryAsync(Me.rozettaApiUser, Me.translateId)
            task.Wait(Interval)
            Dim translate As Translate = task.Result
            If translate Is DirectCast(Nothing, Object) Then Return

            For Each item As TranslateItem In translate.items

                If Not item.done Then
                    Return
                End If
            Next

            Dim ids As List(Of String) = New List(Of String)()

            For Each item As TranslateItem In translate.items
                ids.Add(item.translateItemId)
            Next

            Dim strIds As String = Newtonsoft.Json.JsonConvert.SerializeObject(ids)
            Dim downloadTask As Task(Of Boolean) = Me.client.DownloadZipAsync(Me.rozettaApiUser, strIds, "C:\mydocuments")
            downloadTask.Wait(Interval)
            Me.done = downloadTask.Result
            If Me.done Then Me.StopTimer(Me.aTimer)
        End Sub

        Private Sub StopTimer(ByVal timer As Timers.Timer)
            Me.aTimer.Enabled = False
            Me.aTimer.[Stop]()
            Me.aTimer.Dispose()
        End Sub
    End Class
End Namespace
